### ISSUE TITLE
Calling constructors from within the java script

### ISSUE THREAD

**Issue by bhickey777** [LiquidPlayer/LiquidCore#15]
Is it possible to call constructors from within the javascript. For Example if I have a PageDefinition which extends JSObject with the supporting JSContext constructor, should I be able to do the following in javascript: "var createPageDef = function() {  var pageDefinition = PageDefinition(\"My Page Def\"); return pageDefinition; }; ");

Thx, appreciate the assist

**Comment by ericwlange** [comment]
Yes.  But you would extend the `JSFunction` object, not `JSObject`.  For example:

```
public class PageDefinition extends JSFunction {
    public PageDefinition(JSContext ctx) {
        super(ctx, "constructor"); // <-- This will set up your constructor
    }

    public void constructor(String string) {
        // getThis() will return the newly created instance object if this was called with
        // 'new', otherwise it will return whatever the value of 'this' is when the function was
        // called.  So in this case, we are just setting the property 'some_string' on your
        // new JSObject instance
        getThis().property("some_string", string);
    }
}
```

Note, in your original question, you wrote `var pageDefinition = PageDefinition("My Page Def");`.  This will call the function as a function, not as a constructor.  So, `getThis()` will be whatever the current context of `this` is from the call.  You probably meant `var pageDefinition = *new* PageDefinition("My Page Def");` which will call it as a constructor, and `getThis()` will return the newly constructed instance object.

**Comment by bhickey777** [comment]
Thx, appreciate the assist.

**Comment by ayunawani** [comment]
Hi,

In the above example you have set "some_string " as a property of js object created by "constructor".
`getThis().property("some_string", string);`

Similarly, I want to set a java function as a property of object created by "constructor". 
`getThis().property("some_function", myfunction);`

How can i achieve that?

Thanks,
Ayush Nawani

**Comment by ericwlange** [comment]
```java
JSFunction myFunc = new JSFunction(jsContext, "myFunc") {
    public String myFunc(String a, Integer b, Boolean c) {
        if (c == null || !c) return ":-(";
        String out = (a != null) ? a : "";
        if (b != null) out += ": " + b;
        return out;
    }
});
getThis().property("some_function", myFunc);
```

Then, in Javascript:
```javascript
var pageDefinition = new PageDefinition("my Page Def"));
console.log(pageDefinition.some_function('Foo', 3, true));
console.log(pageDefinition.some_function('Bar',3, 0));
console.log(pageDefinition.some_function('FooBar'));
```

Assuming you have implemented a `console.log()`, it would print:
```
Foo: 3
:-(
:-(
```

**Comment by ayunawani** [comment]
Thank you Eric for your quick response and example, It is working great. 
Can we also wrap a java class method into a JSFunction and set it as a property of object.

Something like this:
```
  final Method test = Demo.class.getMethod("test", String.class);
  JSFunction testing = new JSFunction(getContext(),test);
  getThis().property("some_function", testing);
```

Thanks,
Ayush

**Comment by ericwlange** [comment]
You can.  Take a look at the various constructors of [JSFunction](https://liquidplayer.github.io/LiquidCoreAndroid/org/liquidplayer/javascript/JSFunction.html) and [JSObject](https://liquidplayer.github.io/LiquidCoreAndroid/org/liquidplayer/javascript/JSObject.html).  There are a couple of less clunky ways to do the same.  The first using the `@jsexport` attribute:

```java
public class MyJSObject extends JSObject {
    public MyJSObject(JSContext ctx) {
        super(ctx);
    }

    @jsexport
    Integer incr(Integer x) {
        return x+1;
    }
}

// ... and then:
JSValue foo = new MyJSObject(jsContext);
jsContext.property("foo", foo);
```

Then you can call `foo.incr()` from Javascript.  Of course you can add as many functions as you want to the object this way.

The other way uses an interface:

```java
public interface MyJSFunctions {
    Integer incr(Integer x);
}

public class MyJSObject extends JSObject implements MyJSFunctions {
    public MyJSObject(JSContext ctx) {
        super(ctx, MyJSFunctions.class);
    }

    @Override
    Integer incr(Integer x) {
        return x+1;
    }    
}
```

You can use it the same way in Javascript.

**Comment by ilyamuromets** [comment]
Is it possible to create new instance with java fields?

**Comment by ericwlange** [comment]
You are doing it correctly.  This is a bug.  I am staring [at the code](https://github.com/LiquidPlayer/LiquidCore/blob/master/LiquidCoreAndroid/src/main/java/org/liquidplayer/javascript/JSFunction.java#L569) where this doesn't do the right thing.  After calling the default constructor, I should also call [`addJSExports()`](https://github.com/LiquidPlayer/LiquidCore/blob/master/LiquidCoreAndroid/src/main/java/org/liquidplayer/javascript/JSFunction.java#L569), but neglected to do so.  Or probably better for me to call it in the `JSObject` default constructor.

You can work around this by calling `addJSExports()` in your `PropertyTest()` default constructor.

**Comment by ilyamuromets** [comment]
After calling **addJSExports()** in default constructor after **super()** I get an error:
`java.lang.reflect.InvocationTargetException; Did you remember to call super?`

**Comment by ericwlange** [comment]
I see why that might be.  `addJSExports()` needs to be called after `context` and `valueRef` are set, but this way it is being called beforehand.  I'm not seeing a good workaround at the moment.  It has to be fixed in `JSFunction.h`, unfortunately.

**Comment by ilyamuromets** [comment]
I have found workaround - reflection:
```
public class PropertyTest extends JSFunction
{
	@jsexport(type = Integer.class)
	public Property<Integer> testField;

	public PropertyTest()
	{
		super();
	}

	public PropertyTest(JSContext ctx)
	{
		super(ctx, "ctor", PropertyTest.class);
	}

	public void ctor(String string)
	{
		getThis().property("some_string", string);

		try
		{
			Method method = JSObject.class.getDeclaredMethod("addJSExports");
			method.setAccessible(true);
			method.invoke(getThis());
		}
		catch (NoSuchMethodException e)
		{
			e.printStackTrace();
		}
		catch (InvocationTargetException e)
		{
			e.printStackTrace();
		}
		catch (IllegalAccessException e)
		{
			e.printStackTrace();
		}
	}

	@jsexport
	public void log()
	{
		int y = 0;
	}
}
```

**Comment by ericwlange** [comment]
Oh yeah, you can call it in the ctor. Of course. You shouldn't need
reflections. It is a protected method. You should just be able to call
`getThis().addJSExports()`. Have you tried that?

On Jul 17, 2017 5:53 PM, "ilyamuromets" <notifications@github.com> wrote:

I have found workaround - reflection:

public class PropertyTest extends JSFunction
{
	@jsexport(type = Integer.class)
	public Property<Integer> testField;

	public PropertyTest()
	{
		super();
	}

	public PropertyTest(JSContext ctx)
	{
		super(ctx, "ctor", PropertyTest.class);
	}

	public void ctor(String string)
	{
		getThis().property("some_string", string);

		try
		{
			Method method = JSObject.class.getDeclaredMethod("addJSExports");
			method.setAccessible(true);
			method.invoke(getThis());
		}
		catch (NoSuchMethodException e)
		{
			e.printStackTrace();
		}
		catch (InvocationTargetException e)
		{
			e.printStackTrace();
		}
		catch (IllegalAccessException e)
		{
			e.printStackTrace();
		}
	}

	@jsexport
	public void log()
	{
		int y = 0;
	}
}

—
You are receiving this because you modified the open/close state.

Reply to this email directly, view it on GitHub
<https://github.com/LiquidPlayer/LiquidCore/issues/15#issuecomment-315783480>,
or mute the thread

**Comment by ilyamuromets** [comment]
IDE highlights when I insert this code in my **ctor** `getThis().addJSExports()`:
'addJSExports()' has protected access in 'org.liquidplayer.javascript.JSObject'

so I'm using reflection.

**Comment by ericwlange** [comment]
Fixed in 0.3.0
